x/events/events.go

```diff
@@ -0,0 +1,49 @@
+// Copyright © 2023 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package events
+
+import (
+	"context"
+
+	"github.com/gofrs/uuid"
+	otelattr "go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/ory/kratos/selfservice/flow"
+	"github.com/ory/x/otelx/semconv"
+)
+
+const (
+	SessionIssued      semconv.Event = "SessionIssued"
+	SessionRevoked     semconv.Event = "SessionRevoked"
+	RegistrationFailed semconv.Event = "RegistrationFailed"
+	IdentityCreated    semconv.Event = "IdentityCreated"
+	LoginFailed        semconv.Event = "LoginFailed"
+)
+
+const (
+	attributeKeyFlowType  semconv.AttributeKey = "FlowType"
+	attributeKeySessionID semconv.AttributeKey = "SessionID"
+)
+
+func AttrFlowType(val flow.Type) otelattr.KeyValue {
+	return otelattr.String(attributeKeyFlowType.String(), string(val))
+}
+
+func AttrSessionID(val uuid.UUID) otelattr.KeyValue {
+	return otelattr.String(attributeKeySessionID.String(), val.String())
+}
+
+// Emit adds an event to the current span in the context.
+func Emit(ctx context.Context, event semconv.Event, opt ...otelattr.KeyValue) {
```

Comment:
Please don't :) Just use the OpenTelemetry API.

Reply:
I like the (partial) abstraction from OpenTelemetry which gives us the flexibility to replace the actual implementation.
Also the code gets really noisy and cluttered when using the OpenTelemetry API directly.
```go
	events.Emit(r.Context(), events.IdentityCreated,
		semconv.AttrIdentityID(i.ID),
		events.AttrFlowType(a.Type))
```
vs
```go
	trace.SpanFromContext(r.Context()).AddEvent(
		events.IdentityCreated.String(),
		trace.WithAttributes(
			append(
				semconv.AttributesFromContext(ctx),
				semconv.AttrIdentityID(i.ID),
				events.AttrFlowType(a.Type),
			)...,
		),
	)
```
I'd much prefer the former.

Reply:
We already have the "flexibility to replace the actual implementation": see https://github.com/ory-corp/cloud/blob/dfe42f3e344f62495253212b117bde14c24e4020/cloudlib/analytics/tracer.go#L70.

`events.Emit` only introduces a layer of obfuscation around the fact that this is a OpenTelemetry span event, a well-known entity with a widely-used API.

Please don't wrap the OpenTelemetry API.

Reply:
I get what @alnr is saying, but the default syntax is really verbose and IMO hard to read if you're not familiar with the otelx package. So I'd prefer the indirection through the Emit function.

Reply:
But maybe we can move it to ory/x?

Reply:
I have to side here with @alnr. The syntax is not so much different except for an append and a `semconv.AttributesFromContext` that we save. It would probably make more sense to have a helper for the attributes instead, for example.

In the end this triggers an otel event, so i think it should be clear that this is otel

Reply:
I disagree with the argumentation because it is not just reducing clutter, but instead does an abstraction. One could, when writing this by hand, easily forget to add all attributes from the context. Nothing prevents you to write

```go
	trace.SpanFromContext(r.Context()).AddEvent(
		events.IdentityCreated.String(),
		trace.WithAttributes(
			semconv.AttrIdentityID(i.ID),
			events.AttrFlowType(a.Type),
		),
	)
```

instead of the code above. And I am pretty sure no one will notice that during review, as it looks totally legit.

In the end everyone will just copy-paste this whole code block and potentially forget to change the event type. That's what happens a lot with span names. If the API would be easy enough to just type it, and not require you to copy-paste because it is too much to remember, less errors will be made.

Reply:
I wrote the function exactly for this reason, because I copy-pasted the code around and found that there was a common usage pattern that can be extracted into a simple function.

I follow this pattern a lot, actually. When I see a repetitive pattern popping up in (usually around three or more) places, I try to think about an abstraction. A wrapper function is really just about the easiest and most efficient abstraction. There are definitely abstractions that add a layer of obfuscation to the code (callbacks, because you can't step through them as easily in the debugger, or interfaces where you can't statically know which implementation is called), but *functions* are just about as cost-free as it gets. If at all, we should break our code more into smaller functions (and name and document them well).

In my opinion, this code is fine.

Reply:
The `events.Emit` API has literally no connection to the OpenTelemetry API anymore, beyond using the slightly awkward `otel/attribute` package for no reason. Why use a custom OpenTelemetry tracer at that point at all?

The point of using the OpenTelemetry API for event emission is precisely to avoid implementing a homegrown leaky abstraction/event emission API and instead use an industry standard instead.

If we feel that the OpenTelemetry API is not a good fit after all, that's fine. But then please completely replace the mechanism instead of wrapping the existing API.

Reply:
Recommended reading: https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction

Reply:
I refactored this a bit and would consider this discussion resolved